src/main/java/frc/robot/subsystems/drive/SwerveModule.java

```diff
@@ -0,0 +1,166 @@
+        _driveSpeedSetpoint  = MathUtil.clamp(moduleCommand.getMagnitude(), -1, 1);
+        _rotateSetpoint = moduleCommand.getHeading();
+
+        double currentHeading = getHeading();
+
+        _rotatePID.setSetpoint(_rotateSetpoint, currentHeading);
+
+        double rotateSpeed = _rotatePID.calculate(currentHeading);// calling setSetpoint() then calculate() causes previous error to always be the same as the current error, meaning the Derivative coefficient will do nothing
+
+        double driveSpeed = _driveSpeedSetpoint * Math.cos(Math.toRadians(currentHeading - _rotateSetpoint));
+
+        if (_rotatePID.atSetpoint())
+        {
+            _rotateMotor.stopMotor();
+        }
+        else
+        {
+            //_rotateMotor.setVoltage(0.2 * 12);
+            _rotateMotor.setVoltage(rotateSpeed * 12);// setVoltage(speed * 12) is similar to set(speed); counteracts battery drain
+        }
+
+        _driveMotor.set(TalonFXControlMode.PercentOutput, driveSpeed);
+        //System.out.println(String.format("rotate Setpoint: %6.2f, current heading: %6.2f", _rotateSetpoint, currentHeading));
+        //System.out.println(String.format("Heading: %6.2f, PID Setpoint: %6.2f, PID Error: %6.2f, PID Output: %6.2f", getHeading(), _rotateSetpoint, _rotatePID.getError(), rotateSpeed));
+    }
+
+    public double getHeading()
+    {
+        return Math.IEEEremainder(-_rotateSensor.get() - _rotationZero, 360);
+    }
+
+    /**
+     * set the zero point so that the module is now facing the "reset offset" direction
+     */
+    public void zeroEncoder()
+    {
+        _rotationZero = Math.IEEEremainder(-_rotateSensor.get() - _resetOffset, 360);
+        System.out.println(_rotationZero);
+    }
+
+    /**
+     * Get the rotation encoder's zero offset 
+     * @return the offset as previously set in degrees
+     */
+    public double getRelativeZero()
+    {
+        return _rotationZero;
+    }
+
+    /**
+     * Set the zero point to align the module to offset encoder inputs
+     * @param zero new zero position; scaled to the range of 0 to 360 degrees
+     */
+    public void setRotationZero(double zero)
+    {
+        _rotationZero = Math.IEEEremainder(zero, 360);
+    }
+
+    /**
+     * get the current position of the drive motor's encoder
+     * @return the position (distance), scaled to a usable unit
+     */
+    public double getDrivePosition()
+    {
+        //return _driveMotor.getEncoder().getPosition();
+        return _driveMotor.getSelectedSensorPosition() * Constants.Drive.DRIVE_ENCODER_TO_DISTANCE;
+    }
+
+    /**
+     * Get the change of position of the module 
+     * @return Vector representing the distance and direction of the module's motion
+     */
+    public Vector updateOdometry()
+    {
+        
+        Vector change = new Vector();
+
+        double newDrivePosition = getDrivePosition();
+
+        change.setMagnitude((newDrivePosition - _drivePosition));
+        change.setHeading(getHeading());
```

Comment:
Can we combine these two lines into a `setPolarPosition` function call?

```suggestion
        change.setPolarPosition((newDrivePosition - _drivePosition), getHeading());
```